src/platform/android/java/chip/platform/NsdManagerServiceResolver.java

```diff
@@ -105,62 +108,180 @@ public void run() {
+                  nsdManagerResolverAvailState.signalFree();
+                }
               }
+              mainThreadHandler.removeCallbacks(timeoutRunnable);
             }
-            mainThreadHandler.removeCallbacks(timeoutRunnable);
-          }
 
-          @Override
-          public void onServiceResolved(NsdServiceInfo serviceInfo) {
-            Log.i(
-                TAG,
-                "Resolved service '"
-                    + serviceInfo.getServiceName()
-                    + "' to "
-                    + serviceInfo.getHost());
-            // TODO: Find out if DNS-SD results for Android should contain interface ID
-            chipMdnsCallback.handleServiceResolve(
-                instanceName,
-                serviceType,
-                serviceInfo.getHost().getHostName(),
-                serviceInfo.getHost().getHostAddress(),
-                serviceInfo.getPort(),
-                serviceInfo.getAttributes(),
-                callbackHandle,
-                contextHandle);
+            @Override
+            public void onServiceResolved(NsdServiceInfo serviceInfo) {
+              Log.i(
+                  TAG,
+                  "Resolved service '"
+                      + serviceInfo.getServiceName()
+                      + "' to "
+                      + serviceInfo.getHost());
+              // TODO: Find out if DNS-SD results for Android should contain interface ID
+              chipMdnsCallback.handleServiceResolve(
+                  serviceInfo.getServiceName(),
+                  serviceInfo.getServiceType(),
+                  serviceInfo.getHost().getHostName(),
+                  serviceInfo.getHost().getHostAddress(),
+                  serviceInfo.getPort(),
+                  serviceInfo.getAttributes(),
+                  callbackHandle,
+                  contextHandle);
 
-            if (multicastLock.isHeld()) {
-              multicastLock.release();
+              if (multicastLock.isHeld()) {
+                multicastLock.release();
 
-              if (nsdManagerResolverAvailState != null) {
-                nsdManagerResolverAvailState.signalFree();
+                if (nsdManagerResolverAvailState != null) {
+                  nsdManagerResolverAvailState.signalFree();
+                }
               }
+              mainThreadHandler.removeCallbacks(timeoutRunnable);
             }
-            mainThreadHandler.removeCallbacks(timeoutRunnable);
-          }
-        });
+          });
+    }
 
-    mainThreadHandler.postDelayed(timeoutRunnable, RESOLVE_SERVICE_TIMEOUT);
+    @Override
+    public void onDiscoveryStarted(String regType) {
+      Log.d(TAG, "Service discovery started. regType: " + regType);
+    }
+
+    @Override
+    public void onServiceLost(NsdServiceInfo service) {
+      // When the network service is no longer available.
+      // Internal bookkeeping code goes here.
+      Log.e(TAG, "Service lost: " + service);
```

Comment:
do you need to call `nsdManager.stopServiceDiscovery(this)` here?

Reply:
No. This only means that one of the DNS-SD services the NSDManager had discovered previously has been lost which can happen most prominently if the service has gone off the network. We should continue trying to discover the services here.